src/ModelContextProtocol/Client/DelegatingChannelReader.cs

```diff
@@ -0,0 +1,155 @@
+    public override bool TryPeek(out T item)
+    {
+        try
+        {
+            return GetReader().TryPeek(out item!);
+        }
+        catch (InvalidOperationException)
+        {
+            item = default!;
+            return false;
+        }
+    }
+
+    /// <inheritdoc/>
+    public override bool TryRead(out T item)
+    {
+        try
+        {
+            return GetReader().TryRead(out item!);
+        }
+        catch (InvalidOperationException)
+        {
+            item = default!;
+            return false;
+        }
+    }
+
+    /// <inheritdoc/>
+    public override ValueTask<bool> WaitToReadAsync(CancellationToken cancellationToken = default)
+    {
+        // First wait for the connection to be established
+        if (!_connectionEstablished.Task.IsCompletedSuccessfully)
+        {
+            return new ValueTask<bool>(WaitForConnectionAndThenReadAsync(cancellationToken));
+        }
+
+        // Then delegate to the active reader
+        return GetReader().WaitToReadAsync(cancellationToken);
+    }
+
+    private async Task<bool> WaitForConnectionAndThenReadAsync(CancellationToken cancellationToken)
+    {
+        await _connectionEstablished.Task.ConfigureAwait(false);
+        return await GetReader().WaitToReadAsync(cancellationToken).ConfigureAwait(false);
+    }
+
+    /// <inheritdoc/>
+    public override ValueTask<T> ReadAsync(CancellationToken cancellationToken = default)
+    {
+        // First wait for the connection to be established
+        if (!_connectionEstablished.Task.IsCompletedSuccessfully)
+        {
+            return new ValueTask<T>(WaitForConnectionAndThenGetItemAsync(cancellationToken));
+        }
+
+        // Then delegate to the active reader
+        return GetReader().ReadAsync(cancellationToken);
+    }
+
+    private async Task<T> WaitForConnectionAndThenGetItemAsync(CancellationToken cancellationToken)
+    {
+        await _connectionEstablished.Task.ConfigureAwait(false);
+        return await GetReader().ReadAsync(cancellationToken).ConfigureAwait(false);
+    }
+
+#if NETSTANDARD2_0
+    public IAsyncEnumerable<T> ReadAllAsync(CancellationToken cancellationToken = default)
+    {
+        return ReadAllAsyncInternal(cancellationToken);
+    }
+#else
+    /// <inheritdoc/>
+    public override IAsyncEnumerable<T> ReadAllAsync(CancellationToken cancellationToken = default)
+    {
+        return ReadAllAsyncInternal(cancellationToken);
+    }
+#endif
+
+#if NETSTANDARD2_0
+    private async IAsyncEnumerable<T> ReadAllAsyncInternal(CancellationToken cancellationToken)
+#else
+    private async IAsyncEnumerable<T> ReadAllAsyncInternal([EnumeratorCancellation] CancellationToken cancellationToken)
+#endif
+    {
+        await _connectionEstablished.Task.ConfigureAwait(false);
+        
+        await foreach (var item in GetReader().ReadAllAsync(cancellationToken).ConfigureAwait(false))
+        {
+            yield return item;
+        }
+    }
```

Comment:
Delete all this. The default implementation if ReadAllAsync should be fine for our purposes since it calls into our other methods.